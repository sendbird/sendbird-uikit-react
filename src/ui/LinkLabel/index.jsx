import React from 'react';
import PropTypes from 'prop-types';

import Label, { LabelTypography, LabelColors } from '../Label';
import { changeColorToClassName } from '../Label/utils';
import './index.scss';

const http = /https?:\/\//;

export default function LinkLabel({
  className,
  src,
  type,
  color,
  children,
}) {
  const url = (http.test(src)) ? src : `http://${src}`;

  return (
    <a
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-link-label',
        color ? changeColorToClassName(color) : '',
      ].join(' ')}
      href={url}
      target="_blank"
      rel="noopener noreferrer"
    >
      <Label
        className="sendbird-link-label__label"
        type={type}
        color={color}
      >
        {children}
      </Label>
    </a>
  );
}

LinkLabel.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  src: PropTypes.string.isRequired,
  type: PropTypes.oneOf(Object.keys(LabelTypography)).isRequired,
  color: PropTypes.oneOf(Object.keys(LabelColors)).isRequired,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
};

LinkLabel.defaultProps = {
  className: '',
};

export const LinkLabelTypography = LabelTypography;
export const LinkLabelColors = LabelColors;
