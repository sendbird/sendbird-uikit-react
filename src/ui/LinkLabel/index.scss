@import '../../styles/variables';

.sendbird-link-label {
  text-decoration: unset;
  &:hover {
    cursor: pointer;
    text-decoration: underline;
  }
  &:focus {
    outline: none;
  }

  &__label {
    display: inline;
  }
}

// colors
.sendbird-link-label .sendbird-label--color-onbackground-1 {
  @include themed() {
    color: t(on-bg-1);
  }
}

.sendbird-link-label .sendbird-label--color-onbackground-2 {
  @include themed() {
    color: t(on-bg-2);
  }
}

.sendbird-link-label .sendbird-label--color-onbackground-3 {
  @include themed() {
    color: t(on-bg-3);
  }
}

.sendbird-link-label .sendbird-label--color-oncontent-1 {
  @include themed() {
    color: t(on-content-1);
  }
}

.sendbird-link-label .sendbird-label--color-oncontent-2 {
  @include themed() {
    color: t(on-content-2);
  }
}

.sendbird-link-label .sendbird-label--color-primary {
  @include themed() {
    color: t(primary-3);
  }
}

.sendbird-link-label .sendbird-label--color-error {
  @include themed() {
    color: t(error--3-2);
  }
}

.sendbird-link-label .sendbird-label--color-secondary-3 {
  @include themed() {
    color: t(secondary-3);
  }
}

