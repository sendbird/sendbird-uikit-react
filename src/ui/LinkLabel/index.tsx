import React, { ReactNode } from 'react';

import Label, { LabelColors, LabelTypography } from '../Label';
import { changeColorToClassName } from '../Label/utils';
import './index.scss';
import { ObjectValues } from '../../utils/typeHelpers/objectValues';
import { openURL } from '../../utils/utils';

const http = /https?:\/\//;

type LinkLabelProps = {
  src: string;
  className?: string | string[];
  type?: ObjectValues<typeof LabelTypography>;
  color?: ObjectValues<typeof LabelColors>;
  children: ReactNode;
};

export default function LinkLabel({ className = '', src, type, color, children }: LinkLabelProps) {
  const url = http.test(src) ? src : `http://${src}`;

  return (
    <a
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-link-label',
        changeColorToClassName(color),
      ].join(' ')}
      href={url}
      target="_blank"
      rel="noopener noreferrer"
      // for mobile
      onTouchEnd={(e) => {
        e.preventDefault();
        e.nativeEvent.stopImmediatePropagation();
        openURL(url);
      }}
    >
      <Label className="sendbird-link-label__label" type={type} color={color}>
        {children}
      </Label>
    </a>
  );
}

export const LinkLabelTypography = LabelTypography;
export const LinkLabelColors = LabelColors;
