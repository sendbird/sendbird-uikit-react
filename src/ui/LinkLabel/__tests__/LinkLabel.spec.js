import React from 'react';
import { render, screen } from '@testing-library/react';

import LinkLabel, { LinkLabelTypography, LinkLabelColors } from "../index";

const LINK_LABEL = 'sendbird-link-label';

describe('ui/LinkLabel', () => {
  it('should render link label', function() {
    const CLASS_NAME = 'example-class-name';
    const text = 'example-text';
    const { container } = render(
      <LinkLabel
        className={CLASS_NAME}
        src="https://www.sendbird.com"
        color={LinkLabelColors.PRIMARY}
        type={LinkLabelTypography.BODY_1}
      >
        {text}
      </LinkLabel>
    );

    expect(
      container.getElementsByClassName('sendbird-link-label')[0].className
    ).toContain(CLASS_NAME);
    expect(
      container.getElementsByClassName('sendbird-link-label')[0].className
    ).toContain(LINK_LABEL);
    expect(
      container.getElementsByClassName('sendbird-link-label')[0].children[0].className
    ).toContain(`${LINK_LABEL}__label`);
    expect(
      screen.getByText(text).className
    ).toContain(`${LINK_LABEL}__label`);
  });

  it('should do a snapshot test of the LinkLabel DOM', function() {
    const text = 'example-text';
    const { asFragment } = render(
      <LinkLabel
        src="https://www.sendbird.com"
        color={LinkLabelColors.PRIMARY}
        type={LinkLabelTypography.BODY_1}
      >
        {text}
      </LinkLabel>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
