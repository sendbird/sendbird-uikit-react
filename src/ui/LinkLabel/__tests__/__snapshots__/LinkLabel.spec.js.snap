// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/LinkLabel should do a snapshot test of the LinkLabel DOM 1`] = `
<DocumentFragment>
  <a
    class=" sendbird-link-label sendbird-label--color-primary"
    href="https://www.sendbird.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    <span
      class="sendbird-link-label__label sendbird-label sendbird-label--body-1 sendbird-label--color-primary"
    >
      example-text
    </span>
  </a>
</DocumentFragment>
`;
